src/middlewared/middlewared/plugins/filesystem_/acl_linux.py

```diff
@@ -12,82 +13,102 @@ class FilesystemService(Service, ACLBase):
+            return
 
         if not os.path.realpath(path).startswith('/mnt/'):
-            raise CallError(f"Changing permissions on paths outside of /mnt is not permitted: {path}",
-                            errno.EPERM)
+            verrors.add(
+                f'{schema}.path',
+                f"Changing permissions on paths outside of /mnt is not permitted: {path}"
+            )
 
-        if os.path.realpath(path) in [x['path'] for x in self.middleware.call_sync('pool.query')]:
-            raise CallError(f"Changing permissions of root level dataset is not permitted: {path}",
-                            errno.EPERM)
+        elif len(p.resolve().parents) == 2:
+            verrors.add(
+                f'{schema}.path',
+                f'The specified path is a ZFS pool mountpoint "({path})" '
+            )
+
+        elif self.middleware.call_sync('pool.dataset.path_in_locked_datasets', path):
+            verrors.add(
+                f'{schema}.path',
+                'Path component for is currently encrypted and locked'
+            )
 
     def chown(self, job, data):
         job.set_progress(0, 'Preparing to change owner.')
-
-        self._common_perm_path_validate(data['path'])
+        verrors = ValidationErrors()
 
         uid = -1 if data['uid'] is None else data['uid']
         gid = -1 if data['gid'] is None else data['gid']
         options = data['options']
 
+        self._common_perm_path_validate("filesystem.chown",
+                                        data['path'],
+                                        options.get('recursive', False),
+                                        verrors)
+        verrors.check()
+
         if not options['recursive']:
             job.set_progress(100, 'Finished changing owner.')
             os.chown(data['path'], uid, gid)
-        else:
-            if uid == -1 and gid == -1:
-                return
-            job.set_progress(10, f'Recursively changing owner of {data["path"]}.')
-            # TODO: plumb in acltool to handle recursive / traverse so that we
-            # don't break mountpoints
-            # self.acltool(data['path'], 'chown', uid, gid, options)
-            if gid == -1:
-                chown = subprocess.run(['chown', '-R', str(uid), data['path']],
-                                       check=False, capture_output=True)
-            elif uid == -1:
-                chown = subprocess.run(['chgrp', '-R', str(gid), data['path']],
-                                       check=False, capture_output=True)
-            else:
-                chown = subprocess.run(['chown', '-R', f'{uid}:{gid}', data['path']],
-                                       check=False, capture_output=True)
+            return
 
-            if chown.returncode != 0:
-                raise CallError(f"Failed to chown [{data['path']}]: "
-                                f"{chown.stderr.decode()}")
+        if uid == -1 and gid == -1:
```

Comment:
What does this mean in the context of this method?

Reply:
Passing -1 as an id to chown(2) leaves owner / group same. I think it probably would make more sense in this case to raise a validation error since it doesn't make sense to leave both unchanged.